runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java

```diff
@@ -254,17 +268,23 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
         WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
         BoundedWindow timerWindow = windowNamespace.getWindow();
         delegateEvaluator.onTimer(timer, timerWindow);
+        if (timerInternals.containsUpdateForTimeBefore(currentInputWatermark)) {
```

Comment:
So if I understand this correctly, the `PriorityQueue` causes the timers in the incoming bundle to be fired in timestamp order. That's good. And then if any timer is set that would already be eligible for firing, it ends processing and pushes the rest back. I think this is not optimal, because you could add timer updates to the priority queue and keep firing any eligible timer. The bundle only needs to end when there are no more eligible timers. But on the other hand, I do think it fixes the problem, so happy to merge it.

Reply:
The problem here is that I don't want to split the logic. `WatermarkManager` is responsible for handling all timer setting/firing/removing, so I didn't want to intrude into the logic from `StatefulParDoEvaluator`. If the current logic is fine, I'd merge it as it is.